net: Provide block templates to peers on request

ajtowns opened the pull request:
Implements the sending side of `SENDTEMPLATE`, `GETTEMPLATE`, `TEMPLATE` message scheme for sharing block templates with peers via compact block encoding/reconstruction.

ajtowns committed:
node/miner: Log via MINER debug category

This removes "CreateNewBlock" messages from logs by default. Also changes
"-printpriority" argument to automatically enable -debug=miner logging.

ajtowns committed:
net_processing: add support for template sharing

This adds suport for SENDTEMPLATE, GETTEMPLATE, TEMPLATE p2p messages to generate
and send block templates to our peers. It does not support requesting templates from
our peers.

ajtowns committed:
protocol.h: Define template sharing protocol messages

Also covers functional test setup.

ajtowns committed:
node/miner: drop node/context include

ismaelsadeeq commented:
Nice idea.

A couple of questions:

1. How do nodes build the block template they share with peers? When they use only their mempool to build those templates, I don’t think that the transactions with diverging mempool policy will relay through the network and improve the compact block reconstruction process via this mechanism. Only peers whose outbound connections are miners will benefit.

2. We don’t accept transactions in the mempool that have divergent policy because we deem them non-standard. How does this proposal prevent a scenario where an adversary stuff transactions that are harmful to peers, forcing them to validate and save them? (I think this is where weak blocks are superior to this proposal because they show that work has been done previously.)

3. Is there a limit to the memory used to store the transactions that violate a node’s mempool policy but are received from peers? How do you decide which to keep and which to evict when the limit is reached?
One approach could be linearization based on incentive compatibility and then evicting low-fee transactions, although this has the drawback of then making us not saving transactions paid for via out-of-band.

ajtowns committed:
node/miner: allow generation of oversized templates

Constructing a BlockAssembler with ALLOW_OVERSIZED_BLOCKS allows oversized blocks.

ajtowns commented:
> Maybe a stupid question and I'm not understanding something, but what is the point of this if we have to wait for all transactions to be announced? Isn't the whole idea to make sure our peers know about what we think will be the next block?

If you include a transaction in a template you send to a peer, that implicitly announces the tx to that peer.

The reason we delay announcing txs to a peer is to add some uncertainty about when precisely we received a transaction, and the order in which we received similarly recent transactions -- knowing when many nodes in the network first heard about transactions with precision allows you to make a very good guess about how the network is structured and who heard about a transaction first, which makes it easier to work out who created the transaction which is bad for privacy, and possibly identify weak points in the p2p network to better plan DoS attacks.

So preserving the same timing uncertainty when adding a potentially new way to announce txs seems worthwhile.

ajtowns committed:
blockencodings: check for coinbase before prefilling

When the first tx in a block is a coinbase, always prefill it. But
also support skipping prefill when the first tx is not a coinbase,
for template sharing.

ajtowns committed:
rpc/net: provide experimental gettemplateinfo

mzumsande reviewed (commented):
Would be helpful to link the BIP PR/Delving Post in the OP.

I looked through the code, but will leave most low-level comments for later.

On a conceptual level, this seems like a strong fingerprinting vector (tor/clearnet response would be identical if requests happen roughly at the same time and the same template is used).
I'm not sure if we have given up on that, considering that there seems to be no lack of other existing fingerprinting methods.

Crypt-iQ commented on src/net_processing.cpp:
```diff
@@ -195,6 +197,11 @@ static constexpr double MAX_ADDR_RATE_PER_SECOND{0.1};
+/** How frequently to update templates for sharing */
+static constexpr std::chrono::microseconds TEMPLATE_UPDATE_INTERVAL{30s};
+/** Template weight limit */
+static constexpr unsigned int MAX_TEMPLATE_WEIGHT{8000000};
```
Does `MAX_PROTOCOL_MESSAGE_LENGTH` need to change if a peer requests most txns in a > 4MB template?
